rclcpp_action/src/server.cpp

```diff
@@ -12,30 +12,474 @@
+    rcl_action_goal_handle_t * rcl_handle = rcl_action_accept_new_goal(
+      pimpl_->action_server_.get(), &goal_info);
+    if (!rcl_handle) {
+      throw std::runtime_error("Failed to accept new goal\n");
+    }
+
+    std::shared_ptr<rcl_action_goal_handle_t> handle(new rcl_action_goal_handle_t, deleter);
+    // Copy out goal handle since action server storage disappears when it is fini'd
+    *handle = *rcl_handle;
+
+    if (GoalResponse::ACCEPT_AND_EXECUTE == status) {
+      // Change status to executing
+      ret = rcl_action_update_goal_state(handle.get(), GOAL_EVENT_EXECUTE);
+      if (RCL_RET_OK != ret) {
+        rclcpp::exceptions::throw_from_rcl_error(ret);
+      }
+    }
+    // publish status since a goal's state has changed (was accepted or has begun execution)
+    publish_status();
+
+    // Tell user to start executing action
+    call_goal_accepted_callback(pimpl_->action_server_, handle, uuid, message);
+  }
+}
+
+void
+ServerBase::execute_cancel_request_received()
+{
+  rcl_ret_t ret;
+  rmw_request_id_t request_header;
+
+  // Initialize cancel request
+  auto request = std::make_shared<action_msgs::srv::CancelGoal::Request>();
+
+  ret = rcl_action_take_cancel_request(
+    pimpl_->action_server_.get(),
+    &request_header,
+    request.get());
+
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+
+  // Convert c++ message to C message
+  rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
+  for (size_t i = 0; i < 16; ++i) {
+    cancel_request.goal_info.goal_id.uuid[i] = request->goal_info.goal_id.uuid[i];
+  }
+  cancel_request.goal_info.stamp.sec = request->goal_info.stamp.sec;
+  cancel_request.goal_info.stamp.nanosec = request->goal_info.stamp.nanosec;
+
+  // Get a list of goal info that should be attempted to be cancelled
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  ret = rcl_action_process_cancel_request(
+    pimpl_->action_server_.get(),
+    &cancel_request,
+    &cancel_response);
+
+  auto response = std::make_shared<action_msgs::srv::CancelGoal::Response>();
+
+  auto & goals = cancel_response.msg.goals_canceling;
+  // For each canceled goal, call cancel callback
+  for (size_t i = 0; i < goals.size; ++i) {
+    const rcl_action_goal_info_t & goal_info = goals.data[i];
+    std::array<uint8_t, 16> uuid;
+    for (size_t i = 0; i < 16; ++i) {
+      uuid[i] = goal_info.goal_id.uuid[i];
+    }
+    auto response_pair = call_handle_cancel_callback(uuid);
+    if (CancelResponse::ACCEPT == response_pair.first) {
+      ret = rcl_action_update_goal_state(response_pair.second.get(), GOAL_EVENT_CANCEL);
+      if (RCL_RET_OK != ret) {
+        rcl_ret_t fail_ret = rcl_action_cancel_response_fini(&cancel_response);
+        (void)fail_ret;  // TODO(sloretz) do something with error during cleanup
+        rclcpp::exceptions::throw_from_rcl_error(ret);
+      } else {
```

Comment:
Since we're throwing above, we can remove the `else` block (unnecessary indentation),

Reply:
Fixed, i think. I don't see an else condition anymore.